src/main/main.c

```diff
@@ -163,6 +171,135 @@ void listen_to_buttons (void *pvParameters) {
     }
 }
 
+esp_bd_addr_t trusted_device;
+bool paired = false;
+
+static char *bda2str(esp_bd_addr_t bda, char *str, size_t size) {
+    if (bda == NULL || str == NULL || size < 18) {
+        return NULL;
+    }
+
+    sprintf(str, "%02x:%02x:%02x:%02x:%02x:%02x", bda[0], bda[1], bda[2], bda[3], bda[4], bda[5]);
+    return str;
+}
+
+void esp_bt_gap_cb(esp_bt_gap_cb_event_t event, esp_bt_gap_cb_param_t *param) {
+    char bda_str[18] = {0};
+    
+    switch (event) {
+        case ESP_BT_GAP_AUTH_CMPL_EVT: {
+            //printf("Authentication status: %d, name: %s bda:[%s]", param->auth_cmpl.stat, param->auth_cmpl.device_name, bda2str(param->auth_cmpl.bda, bda_str, sizeof(bda_str)));
+            if (param->auth_cmpl.stat == ESP_BT_STATUS_SUCCESS) {
+                memcpy(trusted_device, param->auth_cmpl.bda, sizeof(esp_bd_addr_t));
+            }
+            break;
+        }
+        case ESP_BT_GAP_PIN_REQ_EVT: {
+            printf("ESP_BT_GAP_PIN_REQ_EVT min_16_digit:%d\n", param->pin_req.min_16_digit);
+            if (param->pin_req.min_16_digit) {
+                printf("Input pin code: 0000 0000 0000 0000\n");
+                esp_bt_pin_code_t pin_code = {0};
+                esp_bt_gap_pin_reply(param->pin_req.bda, true, 16, pin_code);
+            } else {
+                printf("Input pin code: 1234\n");
+                esp_bt_pin_code_t pin_code;
+                pin_code[0] = '1';
+                pin_code[1] = '2';
+                pin_code[2] = '3';
+                pin_code[3] = '4';
+                esp_bt_gap_pin_reply(param->pin_req.bda, true, 4, pin_code);
+            }
+            break;
+        }
+        case ESP_BT_GAP_ACL_CONN_CMPL_STAT_EVT: {
+            //printf("Device connection status: %d\n", param->acl_conn_cmpl_stat.stat);
+            if (param->acl_conn_cmpl_stat.stat != ESP_BT_STATUS_HCI_SUCCESS || !paired) {
+                break;
+            }
+
+            bt_connected = true;
+
+            if (is_locked && !unlocking) {
+                unlocking = true;
+                unlock_timer = 0;
+            }
+            break;
+        }
+        case ESP_BT_GAP_ACL_DISCONN_CMPL_STAT_EVT: {
+            //printf("Device disconnected status: %d\n", param->acl_disconn_cmpl_stat.reason);
+            /*if (param->acl_disconn_cmpl_stat.reason == ESP_BT_STATUS_HCI_PEER_USER) {
+                paired = true;
+            }*/ // Works only when device gets to call the disconnect after pairing
+            paired = true; // Works always
```

Comment:
"To ensure the finishing of the handshake between the phone and ESP, the paired has to be true"